### PR TITLE
Implement the equivalent of torch.cat for grad tensors

### PR DESCRIPTION
**Pull request by jla524**
for https://github.com/geohot/tinygrad/issues/274

### CODE REVIEW

**Comment by geohot** [comment]
This only works on the CPU, if you want to add an op, you have to make it work on all natively.

**Comment by jla524** [comment]
Hi George, thanks for reviewing my PR. Can you explain why it only works on CPU? I did some testing with GPU and it seems to be working - am I missing something here?
```
$ GPU=1 ipython3
Python 3.9.5 (default, May  4 2021, 03:36:27) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.24.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from tinygrad.tensor import Tensor

In [2]: x = Tensor.randn(3, 2)

In [3]: x.device
Out[3]: 1

In [4]: y = Tensor.cat((x, x, x))

In [5]: y.device
Out[5]: 1

In [6]: x.cpu().data
Out[6]: 
array([[-1.8184614 , -0.6809074 ],
       [-0.57175547, -0.69426394],
       [ 0.12231188, -0.9141508 ]], dtype=float32)

In [7]: y.cpu().data
Out[7]: 
array([[-1.8184614 , -0.6809074 ],
       [-0.57175547, -0.69426394],
       [ 0.12231188, -0.9141508 ],
       [-1.8184614 , -0.6809074 ],
       [-0.57175547, -0.69426394],
       [ 0.12231188, -0.9141508 ],
       [-1.8184614 , -0.6809074 ],
       [-0.57175547, -0.69426394],
       [ 0.12231188, -0.9141508 ]], dtype=float32)
```

**Comment by geohot** [comment]
It's copying back to the CPU, it needs to be native on the GPU to be a first class op.

**Comment by jla524** [comment]
Thanks for the clarification. I'm not really sure how to implement it without using ".data", are there any examples that I can follow?

**Comment by geohot** [comment]
See ops_cpu.py. It's already ".data" by the time it's passed in there.

**Comment by skoshx** [comment]
@jla524 You also have to also take training into consideration, when adding ops. So there needs to be a `backward` function, so that you can backpropagate properly. I recommend checking out how PyTorch handles it, and then implementing the backwards function based on your findings :) I am happy to help with this PR, since `Cat` is used quite a bit by YOLO for instance, so just ask, and I will try to answer ASAP 👍

**Comment by jla524** [comment]
Hi @skoshx, thanks for reaching out to me! Sorry for the stupid question, but I wanted to know what's wrong with my new approach.

I've added this code in ops_cpu.py:
```
class Cat(Function):                                                                                                                    
  def forward(ctx, arrays, axis=0):                                                                                                     
    shapes = [a.shape for a in arrays]                                                                                                  
    ctx.save_for_backward(shapes)                                                                                                       
    return np.concatenate(arrays, axis=axis)                                                                                            
                                                                                                                                        
  def backward(ctx, x):                                                                                                                 
    pass 
 ```

And it keeps giving me this error when I try to call it from tensor
```
$ ipython
Python 3.9.5 (default, May  4 2021, 03:36:27) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.24.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from tinygrad.tensor import Tensor

In [2]: x = Tensor.randn(3, 2)

In [3]: c = Tensor.cat((x, x))
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-3-8fcd21341317> in <module>
----> 1 c = Tensor.cat((x, x))

~/Documents/projects/contrib/tinygrad/tinygrad/tensor.py in dispatch(*x, **kwargs)
    338   Tensor.ops[device][name] = fxn
    339   def dispatch(*x, **kwargs):
--> 340     tt = [arg for arg in x if isinstance(arg, Tensor)][0]
    341     x = [Tensor(np.array([arg], dtype=tt.dtype), device=tt.device, requires_grad=False) if not isinstance(arg, Tensor) else arg for arg in x]
    342     f = Tensor.ops[tt.device][name]

IndexError: list index out of range
```

**Comment by skoshx** [comment]
@jla524 First class ops are accessible from instances of `Tensor` class, not the `Tensor` class itself. So to do what you want would be something like this:
```python
x = Tensor.randn(3, 2)
x = x.cat([x])
```

You might want to check if arrays is an array, or a Tensor, since then you could do `x.cat(y)` & `x.cat((x, y, z))`

**Comment by jla524** [comment]
Hi @skoshx, so I sort of figured out how to implement the operation on CPU, but I'm having a really hard time with the GPU (specially when number of dimensions > 2). I found an example from [tensorflow](https://github.com/tensorflow/tensorflow/blob/c0d87856d5b57d81b29f28141f739becd258a0f5/tensorflow/core/kernels/concat_op.cc) but am unsure what the code does. Please let me know if you have any ideas or suggestions to make the operation work on GPU, thank you!

**Comment by skoshx** [comment]
@jla524 Yes, everything is so easy in Numpy land, but implementing these algorithms by hand is a completely different thing. I recommend you reference [Caffe's cuda implementation](https://github.com/BVLC/caffe/blob/master/src/caffe/layers/concat_layer.cu#L28-L46), their [CPU implementation](https://github.com/BVLC/caffe/blob/master/src/caffe/layers/concat_layer.cpp) is also quite instructive, and can help you understand how concat works with an arbitrary amount of dimensions.

Basically it iterates the complete size of the input (called `bottom` in caffe land), so if the shape is (2, 3), then it iterates `2 * 3` (`numel()` in PyTorch) and calls the cuda kernel, after which it increments an offset, and repeats.

Im (finally) coming to a finish with implementing strided max & avg pool, so after that I can check this out more thoroughly, but I hope the Caffe code is of use for now! Good luck :)

**Comment by geohot** [comment]
Hmm, this is a another core op. Do we need it?

Write tests for it, and I'll try to write it as a second order op.

**Comment by geohot** [comment]
Sorry, didn't mean to close. Regardless, this op needs tests

**Comment by jla524** [comment]
Added more test cases for this. For now it only works on CPU. I think GPU works if the tensors have <= 2 dimensions -  I couldn't figure out how to make it work for > 2.

**Comment by geohot** [comment]
Thinking about it more, Cat isn't the core op. The core op is something like Copy.

To concat two tensors, first you Slice the initial one, then Copy in the others. Copy would also implement setitem.

**Comment by jla524** [comment]
I rewrote Cat as a second order op by adapting sixChar's code (Slice + Add). Backward pass doesn't work yet.
GPU test seems to fail when we have more dimensions - I think it lacks precision.

**Comment by geohot** [comment]
Took your test, but 18 lines instead of 38.

https://github.com/geohot/tinygrad/commit/3cdc77f526081551730498f6bfcf4158595bd91a
